### PR TITLE
fix flawed dynamic helper architecture

### PR DESCRIPTION
**Pull request by jim-minter**
### Which issue this PR addresses:

1) dynamic helper merge brokenness where slices are not truncated when we expected them to be
2) dynamic helper merge brokenness where omitEmpty fields are silently ignored

### What this PR does / why we need it:

1. Reduce the scope of the dynamic helper
   * Split out GVR resolution functionality into a separate, smaller interface
   * Move all the type-unsafe *unstructured.Unstructed codebase used only by Geneva actions into the frontend/adminactions package
2. Move the remainder of the dynamic helper to use runtime.Objects
3. Reimplement the merge strategy in a way that I think is now sane

### Test plan for issue:

Tested manually

### Is there any documentation that needs to be updated for this PR?

No

### CODE REVIEW

**Comment by mjudeikis** [comment]
Fixes: https://github.com/Azure/ARO-RP/issues/1103

File: pkg/cluster/fixmcsuserdata_test.go [hunk 1]
```diff
@@ -23,7 +23,6 @@ import (
 
 	"github.com/Azure/ARO-RP/pkg/api"
 	"github.com/Azure/ARO-RP/pkg/util/cmp"
-	_ "github.com/Azure/ARO-RP/pkg/util/scheme"
```
Comment (jim-minter) [inline pkg/cluster/fixmcsuserdata_test.go:26]: a few lines like this crept into this PR - they are legitimate and orthogonal

File: pkg/util/dynamichelper/createorder.go [hunk 2]
```diff
@@ -4,40 +4,36 @@ package dynamichelper
...
+var createOrderMap = map[reflect.Type]int{
 	// non-namespaced resources
-	"CustomResourceDefinition.apiextensions.k8s.io":    1, // before custom resources
-	"ClusterRole.rbac.authorization.k8s.io":            2, // before workload resources
-	"ClusterRoleBinding.rbac.authorization.k8s.io":     3, // before workload resources
-	"SecurityContextConstraints.security.openshift.io": 4, // before workload resources
+	reflect.TypeOf(&extensionsv1beta1.CustomResourceDefinition{}): -8, // before custom resources
+	reflect.TypeOf(&rbacv1.ClusterRole{}):                         -7, // before workload resources
+	reflect.TypeOf(&rbacv1.ClusterRoleBinding{}):                  -6, // before workload resources
+	reflect.TypeOf(&securityv1.SecurityContextConstraints{}):      -5, // before workload resources
```
Comment (mjudeikis) [inline pkg/util/dynamichelper/createorder.go:23]: Why you use negative values? want to understand thinking :)

File: pkg/util/dynamichelper/dynamichelper.go [hunk 3]
```diff
@@ -5,289 +5,243 @@ package dynamichelper
...
 		return nil, err
 	}
 
 	return dh, nil
 }
 
-func (dh *dynamicHelper) RefreshAPIResources() error {
-	var cli discovery.DiscoveryInterface
-	cli, err := discovery.NewDiscoveryClientForConfig(dh.restconfig)
+func (dh *dynamicHelper) EnsureDeleted(ctx context.Context, groupKind, namespace, name string) error {
```
Comment (mjudeikis) [inline pkg/util/dynamichelper/dynamichelper.go:66]: Please add comment to this function. Name is bit ambiguous.

**Review by mjudeikis — COMMENTED** [review]
This is not the PR you want to review on monday morning :D

File: pkg/util/dynamichelper/createorder.go (hunk 2 above)
Comment (gvanderpotte) [reply pkg/util/dynamichelper/createorder.go:23]: I guess this is to ensure that any key not in the map returns the zero value for int, which will be greater than any value of an already known type in the map. And hence will be done at the end of the order. This is more "concise" and "elegant" than previous code.
Let's see if I guessed right :)

Comment (jim-minter) [reply pkg/util/dynamichelper/createorder.go:23]: Yes, you did ;-)

**Review by mjudeikis — APPROVED** [review]

**Pull request merged**